rustPlatform.buildRustPackage: build auditable binaries

figsoda opened the pull request:
###### Description of changes

closes #200563 

bootstraps `cargo-auditable` and makes all packages built by `buildRustPackage` auditable unless overridden with `auditable = false`

built `fd` on master and ran `rust-audit-info` on it, it should have no sizable impact on binary size

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

figsoda committed:
rustPlatform.buildRustPackage: build auditable binaries

zowoq commented:
Any other distros doing this?

figsoda commented:
just void right now as far as I'm aware

zowoq commented:
> https://github.com/NixOS/nixpkgs/issues/200563
> Figure out whether we want to make this the default

TBH I think making it the default is a bit premature. Keep it opt-in for testing and revisit making it the default has been in master for a few weeks?

lukegb commented:
Given that the changes to the hook will cause every package to be rebuilt anyway I don't think there's necessarily much harm in enabling it by default now, but I'm also not familiar enough with the Rust ecosystem to know whether or not just doing this is likely to be harmful in some edgecase.

zowoq commented:
> maybe we can opt in for a small set of packages and make it the default when it's observed to be stable?

^ This is what I meant when I said this:

> Keep it opt-in for testing and revisit making it the default has been in master for a few weeks?

___

> Given that the changes to the hook will cause every package to be rebuilt anyway

Don't see how this makes any difference as packages will be rebuilt in the following staging cycles anyway?

> I don't think there's necessarily much harm in enabling it by default now, but I'm also not familiar enough with the Rust ecosystem to know whether or not just doing this is likely to be harmful in some edgecase.

Reverting it after it is in master would mean another staging cycle so I think a change that affects every use of `buildRustPackage` warrants a bit of testing before making it the default?

lukegb commented:
I think my main complaint is that "a bit of testing" does not a testing plan make :)

I agree that it is probably a good idea to do some form of testing, but also... this _is_ nixos-unstable. The purest form of smoke testing here would be "running all the NixOS tests that exist and seeing if they pass", which is a good bet to happen in staging-next, and if there are any obvious problems there it's easy to back out. I agree if it escapes into master then it's a bit more annoying to revert but at least we'll have evidence of real problems at that point.

What sort of shape of testing would you be interested in to help build confidence in this change?

zowoq commented:
> What sort of shape of testing would you be interested in to help build confidence in this change?

This sounds good:

> maybe we can opt in for a small set of packages and make it the default when it's observed to be stable

figsoda committed:
rustPlatform.buildRustPackage: make it not auditable by default

figsoda committed:
alacritty: make auditable

figsoda committed:
rust-analyzer: make auditable

figsoda committed:
jumpy: make auditable

figsoda committed:
fd: make auditable

figsoda committed:
ripgrep: make auditable

figsoda commented:
changed `auditable = false` to the default and made a couple popular programs auditable also added jumpy for some diversity

figsoda committed:
rav1e: make auditable

zowoq commented:
Maybe `mdbook` as well as it's in the `nix` build closure?

figsoda committed:
mdbook: make auditable

zowoq commented on pkgs/tools/video/rav1e/default.nix:
```diff
@@ -13,17 +13,19 @@ in rustPlatform.buildRustPackage rec {
 
   postBuild = ''
-    cargo cbuild --release --frozen --prefix=${placeholder "out"} --target ${rustTargetPlatformSpec}
+    $cargoCommand cbuild --release --frozen --prefix=${placeholder "out"} --target ${rustTargetPlatformSpec}
```
Doesn't need to happen in this PR but any way we could avoid needing to do this? i.e wrap `cargo-auditable` as `cargo` or something like that?

figsoda replied:
How do we stop the wrapper from recursively calling itself? Looks like cargo-auditable just calls `cargo` directly

https://github.com/rust-secure-code/cargo-auditable/blob/6275e86f503d9419b1528aa38af128882a648ba4/cargo-auditable/src/cargo_auditable.rs#L38

figsoda replied:
Maybe we can look at `CARGO_AUDITABLE_ORIG_ARGS`, but I'm not sure if we should rely on that since it is internal and undocumented

figsoda replied:
Another way to do it is just to add something like a `cargoAuditableHook` that sets `RUSTC_WORKSPACE_WRAPPER`

zowoq replied:
> How do we stop the wrapper from recursively calling itself? Looks like cargo-auditable just calls `cargo` directly

Make `cargo` into `cargo-unwrapped`, substitute `cargo` for `cargo-unwrapped` in `cargo-auditable` and then wrap it?

figsoda replied:
> Make `cargo` into `cargo-unwrapped`

Sounds like the most future-proof way of doing it, but it would complicate the bootstrap process a bit.

Due to the complexity of this, I think it is worth a separate issue once this is merged

zowoq replied:
> Due to the complexity of this, I think it is worth a separate issue once this is merged

That's fine but I do think it's something that should be resolved before `auditable` is made the default.

`$cargoCommand` in an "extra" step like it is here isn't too bad but much easier for everyone if people don't need to know about it.

Should also mean that it works without changes for packages that have cmake, make, meson, etc build systems calling `cargo`.

Shnatsel replied:
Hey! `cargo auditable` author here 👋 

I'll see what I can do about providing a robust way of using `cargo auditable` as a drop-in replacement for `cargo.`

Here's the upstream issue so you can track the progress: https://github.com/rust-secure-code/cargo-auditable/issues/89

figsoda replied:
@Shnatsel thanks for all the work on cargo-auditable.
It works! I wasn't able to use an alias since `cargoBuildHook` uses `env`, which doesn't pick up aliases, but a shell script wrapper does the job

jcgruenhage commented:
> Any other distros doing this? Besides void?

Also not aware of other distros, but we've been building any update to Rust packages that use the cargo build style with auditable, and we have not run into problems with it.

Are there any other distros where adopting this would be easy to implement? I've looked at both Arch and Alpine, and they don't abstract away the build commands into a build style or hook it something like that, meaning that adopting that for all packages is a way bigger change than it would be for Void or NixOS.

figsoda committed:
cargo-auditable: 0.5.5 -> unstable-2022-12-07

Diff: https://github.com/rust-secure-code/cargo-auditable/compare/v0.5.5...246468da22d619c816227797fb176c44026c7105

figsoda committed:
cargo-auditable-cargo-wrapper: init

figsoda committed:
librsvg: make auditable

figsoda commented:
I added `cargo-auditable-cargo-wrapper` and based `buildRustPackage` on it. Not sure about the name, also considered `cargo-auditable.passthru.cargoWrapper` but didn't do that for the simplicity of this initial implementation. open to suggestions

Made `libsrvg` auditable as per @Shnatsel's suggestion since now it's possible with the wrapper

It should be simple to add auditable support for `buildRustCrate` as well, but I am not familiar with it so I'll leave it to someone else

Shnatsel commented:
I've released `cargo auditable` v0.6.0 with support for using it as a drop-in replacement for `cargo`, so that you don't have to use a development snapshot.

figsoda committed:
cargo-auditable: unstable-2022-12-07 -> 0.6.0

Shnatsel commented:
To clarify the potential failure modes: If something goes wrong, the build should fail. It is highly unlikely that a build would produce a non-functional binary, since the data injection mechanism is lifted directly from the Rust compiler, so it's well-tested. And it simply emits an additional object file for the linker to process, so it simply reuses through all the usual, well-trodden codepaths.

figsoda commented:
@zowoq is this good to merge?

zowoq commented on pkgs/development/libraries/librsvg/default.nix:
```diff
@@ -15,7 +15,7 @@
 , rustc
 , rust
-, cargo
+, cargo-auditable-cargo-wrapper
```
Add a TODO here as well for when auditable is the default?

figsoda replied:
Currently it only works for `buildRustPackage`, do you think it should just be a override in cargo then?

zowoq replied:
Making it consistent (opt-out) tree-wide would be better but I guess after `buildRustPackage` is covered there isn't that much left.

Leave it for now and we can revisit this when auditable is made the default?

figsoda replied:
sounds good

Mindavi commented:
When auditable is set to true, this causes a regression when cross-compiling rust packages where a cross-compiled rustc and cargo end up in the closure (causing failures since that doesn't work yet):

```
rick@nixos-asus:~/nixpkgs$ nom build -f default.nix pkgsCross.aarch64-multiplatform.ripgrep -L
these 12 derivations will be built:
  /nix/store/zk6as569vzji8979214qig7y0y26isyr-cargo-1.66.1.drv
  /nix/store/4zqkw80n5lnc9lg4hngqqxnwwnhx3gva-rustc-aarch64-unknown-linux-gnu-1.66.1.drv
  /nix/store/7v21vgng6qjpg4rw96x8nz8b87snv7nm-cargo-check-hook.sh.drv
  /nix/store/jahm46k3xf68zkf15gvjgpmlm2jjrpv2-cargo-build-hook.sh.drv
  /nix/store/9rhmncrcsza15fii2f6644n4mbrhkvbf-cargo-aarch64-unknown-linux-gnu-1.66.1.drv
  /nix/store/hl8g9lskjiqb30sdml2acnn8lja88rqf-cargo-build-hook.sh.drv
  /nix/store/zs37bfg93h0879lglasm7fd96c9v54nm-cargo-check-hook.sh.drv
  /nix/store/30wx8md8q7wlrkqln6ngd6cskrbw55d4-cargo-auditable-aarch64-unknown-linux-gnu-0.6.0.drv
  /nix/store/csz9pipmxk9p9nwcmarn4fz3d4b9vans-cargo.drv
  /nix/store/9cj5wpz3k9lna0vcr2rfs0vwhc31fwpw-cargo-auditable-aarch64-unknown-linux-gnu-0.6.0.drv
  /nix/store/18vrjsy9fsljw6nv0xzpdn083zfzm9dm-cargo.drv
  /nix/store/3h6lcrmxl2scm42bvr34dsdjbg3nw9dn-ripgrep-aarch64-unknown-linux-gnu-13.0.0.drv
┏━ Dependency Graph:
┃             ┌─ ⏳︎ cargo-check-hook.sh waiting for 1 ⏵︎
┃             │     ┌─ ⏳︎ cargo-build-hook.sh waiting for 1 ⏵︎
┃             │     ├─ ⏳︎ cargo-check-hook.sh waiting for 1 ⏵︎
┃             │     │  ┌─ ⏵︎ cargo-1.66.1 (buildPhase) ⏱︎ 15s (∅︎ 6m42s)
┃             │     ├─ ⏳︎ rustc-aarch64-unknown-linux-gnu-1.66.1
┃             │  ┌─ ⏳︎ cargo-aarch64-unknown-linux-gnu-1.66.1
┃             ├─ ⏳︎ cargo-build-hook.sh
┃          ┌─ ⏳︎ cargo-auditable-aarch64-unknown-linux-gnu-0.6.0
┃       ┌─ ⏳︎ cargo
┃    ┌─ ⏳︎ cargo-auditable-aarch64-unknown-linux-gnu-0.6.0
┃ ┌─ ⏳︎ cargo
┃ ⏳︎ ripgrep-aarch64-unknown-linux-gnu-13.0.0
┣━━━ Builds           
┗━ ∑︎ ⏵︎ 1 │ ✔︎ 0 │ ⏳︎ 11 │ ⏱︎ 22s^C
```

figsoda commented:
https://github.com/NixOS/nixpkgs/pull/210394
